File: src/main/java/org/jackhuang/watercraft/util/Mods.java
/**
 * Copyright (c) Huang Yuhui, 2014
 * 
 * "WaterCraft" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */

package org.jackhuang.watercraft.util;

import cpw.mods.fml.common.Loader;
import cpw.mods.fml.common.ModAPIManager;
import cpw.mods.fml.common.versioning.ArtifactVersion;
import cpw.mods.fml.common.versioning.VersionParser;

public class Mods {

    public static class IDs {
        public static final String BuildCraftPower = "BuildCraftAPI|power";
        public static final String BuildCraftCore = "BuildCraft|Core";
        public static final String BuildCraftFactory = "BuildCraft|Factory";
        public static final String Factorization = "factorization";
        public static final String IndustrialCraft2 = "IC2";
        public static final String IndustrialCraft2API = "IC2API";
        public static final String Thaumcraft = "Thaumcraft";
        public static final String ThermalExpansion = "ThermalExpansion";
        public static final String RotaryCraft = "RotaryCraft";
        public static final String CoFHAPIEnergy = "CoFHAPI|energy";
        public static final String GregTech = "gregtech";
        public static final String Waila = "Waila";
        public static final String Mekanism = "Mekanism";
        public static final String CraftGuide = "craftguide";
        public static final String ExNihilo = "ExNihilo";
        public static final String Railcraft = "Railcraft";
        public static final String ImmersiveEngineering = "ImmersiveEngineering";
        public static final String MineTweaker3 = "MineTweaker3";
    }

    public static final SimpleMod Factorization = new SimpleMod(
            IDs.Factorization);
    public static final SimpleMod ThermalExpansion = new SimpleMod(
            IDs.ThermalExpansion);
    public static final SimpleMod IndustrialCraft2 = new SimpleMod(
            IDs.IndustrialCraft2);
    public static final SimpleMod MineTweaker3 = new SimpleMod(
            IDs.MineTweaker3);
    public static final SimpleMod BuildCraftCore = new SimpleMod(
            IDs.BuildCraftCore);
    public static final SimpleMod BuildCraftFactory = new SimpleMod(
            IDs.BuildCraftFactory);
    public static final SimpleMod Railcraft = new SimpleMod(
            IDs.Railcraft);
    public static final SimpleMod BuildCraftPower = new SimpleMod(
            IDs.BuildCraftPower);
    public static final SimpleMod CoFHAPIEnergy = new SimpleMod(
            IDs.CoFHAPIEnergy);
    public static final SimpleMod ImmersiveEngineering = new SimpleMod(
            IDs.ImmersiveEngineering);
    public static final SimpleMod Thaumcraft = new SimpleMod(IDs.Thaumcraft);
    public static final SimpleMod GregTech = new SimpleMod(IDs.GregTech);
    public static final SimpleMod Waila = new SimpleMod(IDs.Waila);
    public static final SimpleMod Mekanism = new SimpleMod(IDs.Mekanism);
    public static final SimpleMod CraftGuide = new SimpleMod(IDs.CraftGuide);
    public static final SimpleMod ExNihilo = new SimpleMod(IDs.ExNihilo);

    public static class SimpleMod {
        public String id;
        public boolean isAvailable;

        public SimpleMod(String id) {
            this.id = id;

            ArtifactVersion version = VersionParser.parseVersionReference(id);
            if (Loader.isModLoaded(version.getLabel()))
                isAvailable = version.containsVersion(Loader.instance()
                        .getIndexedModList().get(version.getLabel())
                        .getProcessedVersion());
            else
                isAvailable = ModAPIManager.INSTANCE.hasAPI(version.getLabel());
        }
    }

}

File: src/main/kotlin/waterpower/integration/Mods.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.integration

import net.minecraftforge.fml.common.Loader
import net.minecraftforge.fml.common.ModAPIManager
import net.minecraftforge.fml.common.versioning.VersionParser

class Mod(val id: String) {
    val isAvailable: Boolean by lazy {
        val version = VersionParser.parseVersionReference(id)
        if (Loader.isModLoaded(version.label))
            version.containsVersion(Loader.instance().indexedModList[version.label]?.processedVersion)
        else
            ModAPIManager.INSTANCE.hasAPI(version.label)
    }

    companion object {
        val AppliedEnergistics2 = Mod(IDs.AppliedEnergistics2)
        val BuildCraftCore = Mod(IDs.BuildCraftCore)
        val BuildCraftFactory = Mod(IDs.BuildCraftFactory)
        val BuildCraftPower = Mod(IDs.BuildCraftPower)
        val CoFHAPIEnergy = Mod(IDs.CoFHAPIEnergy)
        val CraftGuide = Mod(IDs.CraftGuide)
        val EnderIO = Mod(IDs.EnderIO)
        val ExNihilo = Mod(IDs.ExNihilo)
        val Factorization = Mod(IDs.Factorization)
        val GregTech = Mod(IDs.GregTech)
        val ImmersiveEngineering = Mod(IDs.ImmersiveEngineering)
        val IndustrialCraft2 = Mod(IDs.IndustrialCraft2)
        val Mekanism = Mod(IDs.Mekanism)
        val CraftTweaker = Mod(IDs.CraftTweaker)
        val RailCraft = Mod(IDs.RailCraft)
        val Thaumcraft = Mod(IDs.Thaumcraft)
        val ThermalExpansion = Mod(IDs.ThermalExpansion)
        val TinkersConstruct = Mod(IDs.TinkersConstruct)
        val Waila = Mod(IDs.Waila)
    }
}

object IDs {
    const val AppliedEnergistics2 = "appliedenergistics2"
    const val BuildCraftCore = "BuildCraft|Core"
    const val BuildCraftFactory = "BuildCraft|Factory"
    const val BuildCraftPower = "BuildCraftAPI|power"
    const val CoFHAPIEnergy = "CoFHAPI|energy"
    const val CraftGuide = "craftguide"
    const val CraftTweaker = "MineTweaker3"
    const val EnderIO = "EnderIO"
    const val ExNihilo = "exnihilo"
    const val Factorization = "factorization"
    const val GregTech = "gregtech"
    const val ImmersiveEngineering = "ImmersiveEngineering"
    const val IndustrialCraft2 = "IC2"
    const val Mekanism = "Mekanism"
    const val RailCraft = "Railcraft"
    const val RotaryCraft = "RotaryCraft"
    const val Thaumcraft = "Thaumcraft"
    const val ThermalExpansion = "ThermalExpansion"
    const val TinkersConstruct = "TConstruct"
    const val Waila = "Waila"
}